File: Sean的leetcode/python/Minimum Height Trees.py
###
# BFS
# Time Complexity: O(n) 
# Space Complexity: O(n)
###
class Solution(object):
    def findMinHeightTrees(self, n, edges):
        """
        :type n: int
        :type edges: List[List[int]]
        :rtype: List[int]
        """
        if not n:
            return []
        if n == 1:
            return [0]
        table = collections.defaultdict(set)
        for edge in edges:
            table[edge[0]].add(edge[1])
            table[edge[1]].add(edge[0])
        q = [k for k, v in table.iteritems() if len(v) == 1]
        num = n
        while num > 2:
            next_q = []
            num -= len(q)
            for nd in q:
                nb = table[nd].pop()
                table[nb].remove(nd)
                if len(table[nb]) == 1:
                    next_q.append(nb)
            q = next_q
        return q 
        

File: MyLeetCode/Leetcode/Minimum Height Trees.cpp
/* Two pointers, BFS
 * start from the leave and go internal until the number of nodes <= 2
 * Time Complexity: O(n)
 */
class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<pair<int, int>>& edges) {
        vector<int> res;
        vector<unordered_set<int>> table(n);
        if (!n) return res;
        if (n == 1) return {0};
        for (auto p : edges) {
            table[p.first].insert(p.second);
            table[p.second].insert(p.first);
        }
        queue<int> q;
        for (int i = 0; i < table.size(); i++) {
            if (table[i].size() == 1) {
                q.push(i);
            }
        }
        while (n > 2) {
            int num = q.size();
            for (int i = 0; i < num; i++) {
                int tmp = q.front();
                q.pop();
                int adj = *(table[tmp].begin());
                table[adj].erase(tmp);
                if (table[adj].size() == 1) {
                    q.push(adj);
                }
            }
            n -= num;
        }
        while (!q.empty()) {
            int tmp = q.front();
            q.pop();
            res.push_back(tmp);
        }
        return res;
    }
};